modules/renter/hostdb/hosttree/hosttree.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	// node is a node in the tree.
+	node struct {
+		parent *node
+		left   *node
+		right  *node
+
+		count int
+		taken bool
+
+		weight types.Currency
+		entry  *HostEntry
+	}
+)
+
+// createNode creates a new node using the provided `parent` and `entry`.
+func createNode(parent *node, entry *HostEntry) *node {
+	return &node{
+		parent: parent,
+		weight: entry.Weight,
+		count:  1,
+
+		taken: true,
+		entry: entry,
+	}
+}
+
+// New creates a new, empty, HostTree.
+func New() *HostTree {
+	return &HostTree{
+		node: &node{
+			count: 1,
+		},
+		hosts: make(map[string]*node),
+	}
+}
+
+// recursiveInsert inserts an entry into the appropriate place in the tree. The
+// running time of recursiveInsert is log(n) in the maximum number of elements
+// that have ever been in the tree.
+func (n *node) recursiveInsert(entry *HostEntry) (nodesAdded int, newnode *node) {
+	// If there is no parent and no children, and the node is not taken, assign
+	// this entry to this node.
+	if n.parent == nil && n.left == nil && n.right == nil && !n.taken {
+		n.entry = entry
+		n.taken = true
+		n.weight = entry.Weight
+		nodesAdded = 0
+		newnode = n
+		return
+	}
+
+	n.weight = n.weight.Add(entry.Weight)
+
+	// If the current node is empty, add the entry but don't increase the
+	// count.
+	if !n.taken {
+		n.taken = true
+		n.entry = entry
+		newnode = n
+		return
+	}
+
+	// Insert the element into the lest populated side.
+	if n.left == nil {
+		n.left = createNode(n, entry)
+		nodesAdded = 1
+		newnode = n.left
+	} else if n.right == nil {
+		n.right = createNode(n, entry)
+		nodesAdded = 1
+		newnode = n.right
+	} else if n.left.count <= n.right.count {
+		nodesAdded, newnode = n.left.recursiveInsert(entry)
+	} else {
+		nodesAdded, newnode = n.right.recursiveInsert(entry)
+	}
+
+	n.count += nodesAdded
+	return
+}
+
+// nodeAtWeight grabs an element in the tree that appears at the given weight.
+// Though the tree has an arbitrary sorting, a sufficiently random weight will
+// pull a random element. The tree is searched through in a post-ordered way.
+func (n *node) nodeAtWeight(weight types.Currency) (*node, error) {
+	// Sanity check - weight must be less than the total weight of the tree.
+	if weight.Cmp(n.weight) > 0 {
+		return nil, ErrWeightTooHeavy
+	}
+
+	// Check if the left or right child should be returned.
+	if n.left != nil {
+		if weight.Cmp(n.left.weight) < 0 {
+			return n.left.nodeAtWeight(weight)
+		}
+		weight = weight.Sub(n.left.weight) // Search from the 0th index of the right side.
+	}
+	if n.right != nil && weight.Cmp(n.right.weight) < 0 {
+		return n.right.nodeAtWeight(weight)
+	}
+
+	// Should we panic here instead?
+	if !n.taken {
+		return nil, ErrNilEntry
+	}
+
+	// Return the root entry.
+	return n, nil
+}
+
+// remove takes a node and removes it from the tree by climbing through the
+// list of parents. remove does not delete nodes.
+func (n *node) remove() {
+	n.weight = n.weight.Sub(n.entry.Weight)
+	n.taken = false
```

Comment:
why are nodes not deleted, but rather marked as "not taken?" Is it an efficiency thing?